jaxopt/_src/tree_util.py

```diff
@@ -102,4 +108,40 @@ def tree_gram(a):
   """
   vmap_left = jax.vmap(tree_vdot, in_axes=(0,None))
   vmap_right = jax.vmap(vmap_left, in_axes=(None,0))
-  return vmap_right(a, a)
\ No newline at end of file
+  return vmap_right(a, a)
+
+
+def tree_inf_norm(v):
+  """Compute infinity norm of pytree."""
+  leaf_inf_norm = tree_map(lambda l: jnp.max(jnp.abs(l)), v)
+  return tree_reduce(jnp.maximum, leaf_inf_norm)
+
+
+def tree_where(cond, a, b):
+  """jnp.where for for trees. Mimic broadcasting semantic of jnp.where."""
+  la = len(tree_leaves(a))
+  lb = len(tree_leaves(b))
+  lc = len(tree_leaves(cond))
+  if la > 1 and lb > 1:
+    return tree_map(lambda c,u,v: jnp.where(c, u, v), cond, a, b)
+  if la > 1 and lb == 1:
+    return tree_map(lambda c,u: jnp.where(c, u, b), cond, a)
+  if la == 1 and lb > 1:
+    return tree_map(lambda c,v: jnp.where(c, a, v), cond, b)
+  return tree_map(lambda c: jnp.where(c, a, b), cond)
+
+
+def tree_negative(tree):
+  """Computes elementwise negation -x."""
+  return tree_scalar_mul(-1, tree)
+
+
+def tree_reciproqual(tree):
+  """Computes elementwise inverse 1/x."""
+  return tree_map(lambda x: 1/x, tree)
```

Comment:
Maybe replace `1/x` with `jnp.reciprocal` just in case JAX does a fancy optimization.